src/little_dict.jl

```diff
@@ -0,0 +1,245 @@
+        return got
+    end
+end
+
+function Base.iterate(dd::LittleDict, ii=1)
+    ii > length(dd.keys) && return nothing
+    return (dd.keys[ii] => dd.vals[ii], ii+1)
+end
+
+function merge(d1::LittleDict, d2::AbstractDict)
+    return merge((x,y)->y, d1, d2)
+end
+
+function merge(
+    combine::Function,
+    d::LittleDict,
+    others::AbstractDict...
+    )
+    K,V = _merge_kvtypes(d, others...)
+    dc = LittleDict{K,V}(d)
+    for d2 in others
+        for (k2,v2) in d2
+            got = get(dc, k2, NotFoundSentinel())
+            if got isa NotFoundSentinel
+                add_new!(dc, k2, v2)
+            else
+                # GOLDPLATE: ideally we would avoid iterating this twice
+                # once for get and once for setindex!
+                dc[k2]=combine(got, v2)
+            end
+        end
+    end
+    return dc
+end
+
+
+Base.empty(dd::LittleDict{K,V}) where {K,V} = LittleDict{K,V}()
+
+######## Methods that all mutable AbstractDict's should implement
+
+function Base.sizehint!(dd::LittleDict, sz)
+    sizehint!(dd.keys, sz)
+    sizehint!(dd.vals,sz)
+    return dd
+end
+
+function add_new!(dd::LittleDict, key, value)
+    # Not found, add to the end
+    push!(dd.keys, key)
+    try
+        push!(dd.vals, value)
+    catch
+        # if we sucessfully added  a key, but failed to add a value
+        # then we need to remove the key so the little dict remains with constistant state
+        pop!(dd.keys)
+        rethrow()
+    end
+    return dd
+end
+
+
+# This is the safe case that can't error on the adding of values.
+# So no catching exceptions then rollback of adding keys is needed
+# Optimising this case gives a 30% speedup on setindex! when element is new
+# On a vaguely realistic microbenchmark
+function add_new!(dd::LittleDict{<:Any,V,<:Vector,<:Vector{V}}, key, value::V) where {V}
+    # Not found, add to the end
+    push!(dd.keys, key)
+    push!(dd.vals, value)
+    return dd
+end
+
+function Base.setindex!(dd::LittleDict, value, key)
+    # Assertion below commented out as by standards of carefully optimised
+    # setindex! it has huge code (26%), this does mean that if someone has messed
+    # with the fields of the LittleDict directly, then the @inbounds could be invalid
+    #@assert length(dd.keys) == length(dd.vals)
+    @simd for ii in 1:length(dd.keys)
+        cand = @inbounds dd.keys[ii]
+        if isequal(cand, key)
+            @inbounds(dd.vals[ii] = value)
+            return dd
+        end
+    end
+    add_new!(dd, key, value)
+    return dd
+end
+
+function Base.pop!(dd::LittleDict)
+    pop!(dd.keys)
+    return pop!(dd.vals)
+end
+
+function Base.pop!(dd::LittleDict, key)
+    @assert length(dd.keys) == length(dd.vals)
+    
+    @simd for ii in 1:length(dd.keys)
+        cand = @inbounds dd.keys[ii]
+        if isequal(cand, key)
+            deleteat!(dd.keys, ii)
+            val = @inbounds dd.vals[ii]
+            deleteat!(dd.vals, ii)
+            return val
```

Comment:
Bad things happen if you can `pop!` key, but not `pop!` value.
since it will put the dict in an inconsistant state, and will not get an error.
Might might want to put a frozenness check in